### ISSUE TITLE
Unable to input Vietnamese

### ISSUE THREAD

**Issue by superchick** [SpartanJ/ecode#210]
All popular Vietnamese input methods (Unikey, EVKey, OpenKey,...) failed to work.

**Comment by SpartanJ** [comment]
You can't write anything or it just display a rectangle? Because is it's just displaying incorrectly you can fix it by changing the font. I tried by myself and seems to work but I'm really not familiar with Vietnamese.

**Comment by superchick** [comment]
I changed font to Liberation Mono. Still unable to input Vietnamese. For example, with OpenKey, this is what I see when I type the text in https://github.com/SpartanJ/ecode/issues/209:

```
Tra ăm na ăm trong coi õi ngu ưo ươi ười ta
Chu ư ữ tai ài chu ư ữ me ê ệnh kheo éo la à ghet ét nhau
Trai ải qua mo ôt ột cuo ôc uộc be ê ể da âu
Nhu ưng ững d đie êu iều tro ông tha ây ấy ma à d đau d đo ơn ớn lo òng.
```

If copy and paste the text, it displayed correctly. But I want to type new text, not copy and paste existing text.

**Comment by SpartanJ** [comment]
Yeah sorry but I need more information because I have zero idea what could be wrong. What is your OS? You're unable to compose those characters that normally is a combination of other characters, right? Could you send me a video displaying how you normally write on any functioning app and how it looks on ecode, so I can see what's going on? Thanks

**Comment by superchick** [comment]
I will find other code editor.

**Comment by SpartanJ** [comment]
I don't think you understand what's the purpose of reporting bugs on open-source software. Please find any editor you like, but your collaboration would be still appreciated, it would help to improve the experience for everyone. I'll reopen it until we fix it.

**Comment by superchick** [comment]
> I'll reopen it until we fix it.

I don't think you can fix it. You are free to keep it open indefinitely. I don't care.

**Comment by iahung2** [comment]
Don't give up too soon!

**Comment by iahung2** [comment]
@SpartanJ Download this:

https://richdark.github.io/klavier/

Change settings via right clicking the tray icon. Choose to use `K` key instead of `Insert`.

Open Windows' Notepad, type `a` and then press the `K` key repeatedly. You will see the character being changed.

Open ecode, do the same. You will see it will insert new character instead of modifying the existing character you have entered.

**Comment by superchick** [comment]
@SpartanJ Spanish also has accents like Vietnamese. It seems you have never tried to input Spanish in your code editor. Otherwise, you should have discovered this problem already.

p/s: Am I wrong to assume someone from Argentina speaks Spanish?

**Comment by SpartanJ** [comment]
I always write Spanish on my editor and it works perfectly fine on all OSes. But I have no idea how people write Vietnamese.

**Comment by superchick** [comment]
> I changed font to Liberation Mono. Still unable to input Vietnamese. For example, with OpenKey, this is what I see when I type the text in #209:
> 
> ```
> Tra ăm na ăm trong coi õi ngu ưo ươi ười ta
> Chu ư ữ tai ài chu ư ữ me ê ệnh kheo éo la à ghet ét nhau
> Trai ải qua mo ôt ột cuo ôc uộc be ê ể da âu
> Nhu ưng ững d đie êu iều tro ông tha ây ấy ma à d đau d đo ơn ớn lo òng.
> ```
> 
> If copy and paste the text, it displayed correctly. But I want to type new text, not copy and paste existing text.

The space between two parts of a word in the text above is an illusion. It's an artifact created when I copy the text to the browser. What I really see in the code editor is a square. For example, instead of `Nhu ưng ững`, it's `Nhu[a square]ưng[a square]ững`.

**Comment by superchick** [comment]
> I always write Spanish on my editor and it works perfectly fine on all OSes. But I have no idea how people write Vietnamese.

I have no idea how people write Spanish, too. But Google tell me it's kind of like how people input French as mentioned above. Typing Vietnamese is almost the same. For example, for the word `những`, you type `nhuuxng`. You could play with it online here: https://vntyping.com/

**Comment by SpartanJ** [comment]
Thanks for the information. A couple of things: input seems to work fine, your issue is with the text rendering, it seems that from your text example the editor is able to render properly all characters except for the space character. For some reason you are using [this](https://www.compart.com/en/unicode/U+202F) character as an space, instead of the normal / common space character (int 32). And the font does not contain that glyph / character, so it renders a square (when no character is found that's what freetype will give you). So basically we need to provide a more complete fallback font. But I'm really courious why you use that space character and not the common one. You might have other problems but at least that's one of them.

**Comment by SpartanJ** [comment]
It seems that it's not something commonly used and fonts don't usually provide any glyph for that character, so I'll probably add some logic to treat Narrow No-Break Space as a simple space to avoid the issue. Thanks again.

**Comment by superchick** [comment]
This has nothing to do with fonts or text rendering and the input is not fine. I didn't type these invalid space characters. Do you ever think about it? How could I type them? It's your code editor and the input method (OpenKey) failed to work with each other.

**Comment by superchick** [comment]
Take a single character to be more clear.

`u` -> the character `u` in English, nothing new.

`uu` -> will become `ư` character of Vietnamese

`uux` or `ưx` -> will become `ữ` in Vietnamese (a word, but meaningless)

Your code editor will display `u[a square]ư[a square]ữ`. The input is not fine. It's failed.

**Comment by iahung2** [comment]
> You will see it will insert new character instead of modifying the existing character you have entered.

@SpartanJ The key point is here. Btw, I have no idea if it's the input method or `ecode` itself that inserted these `U+202F` that both of you are talking about.

**Comment by iahung2** [comment]
> how people input French as mentioned above

I'm not sure if French people really type like that in real life. I'm not French. Setting up the input method for my language is too involved. It's only the minimal reproducible example I give him to help debug the problem.

**Comment by SpartanJ** [comment]
I think I am able to reproduce some issue, not sure if the same but I installed openkey with telex and while writing random things seems to erase the written text and not replace it with the new or something like that. I'll play around and try to fix it. For some reason the examples don't work even on my browser but I have no idea what I'm doing, at least I can compare browser vs editor typing and try to fix it.

**Comment by superchick** [comment]
If you can, please try with Unikey (https://www.unikey.org/en/), too. Unikey failed in a different way from OpenKey.

**Comment by iahung2** [comment]
With the mention of Unikey, it seems @superchick is apparently using Windows. @SpartanJ is using Linux. @superchick has never said which OS they are using.

**Comment by SpartanJ** [comment]
Sure, thank you! As far as I could investigate this is an SDL2 bug 😞, I need to check if this has been fixed in a more recent version. I was able to reproduce the issue on some other SDL2 applications, but the most popular one, Steam, seems to work a little bit better (not perfect though). I can replicate the weird space character too, it's being sent from SDL2 sadly, this bug might take some time to debug and fix.

@iahung2 yeah, don't worry, seems broken at least in Windows and macOS (I'm testing from macOS, i've not access to a Linux machine this week).

**Comment by SpartanJ** [comment]
> If you can, please try with Unikey (https://www.unikey.org/en/), too. Unikey failed in a different way from OpenKey.

Update:
UniKey is [integrated with macOS](https://www.unikey.org/apple.html) so it's native (you don't need to install anything, just set the input method) and if I use that one it seems to work perfectly fine. OpenKey breaks IME for some reason here too. I won't have access to a Windows system for a week so I'll have to wait, but I'll do my best to fix it ASAP.

**Comment by iahung2** [comment]
> this is an SDL2 bug 😞

https://github.com/SpartanJ/ecode/issues/189#issuecomment-1991823527

**Comment by iahung2** [comment]
> OpenKey breaks IME for some reason here too

Why don't any of you think it's a bug of the input method (OpenKey, Unikey,...)? Both OpenKey and Unikey are open source. @superchick Should you make a bug report to the corresponding developer?

**Comment by superchick** [comment]
> > If you can, please try with Unikey (https://www.unikey.org/en/), too. Unikey failed in a different way from OpenKey.
> 
> Update: UniKey is [integrated with macOS](https://www.unikey.org/apple.html) so it's native (you don't need to install anything, just set the input method) and if I use that one it seems to work perfectly fine. OpenKey breaks IME for some reason here too. I won't have access to a Windows system for a week so I'll have to wait, but I'll do my best to fix it ASAP.

Unikey is both the name of the input method engine and the input method program. Unikey as the input method engine has nothing to do with the Unikey input method program on Windows. On Linux, there are `ibus-unikey` and `fcitx-unikey`, too. They both use the Unikey input method engine but have nothing to do with the Unikey program on Windows.

**Comment by superchick** [comment]
> > OpenKey breaks IME for some reason here too
> 
> Why don't any of you think it's a bug of the input method (OpenKey, Unikey,...)? Both OpenKey and Unikey are open source. @superchick Should you make a bug report to the corresponding developer?

Yes. This should be the first thing I need to do. I don't know why I came here to moan with the ecode developer in the first place. I should contact the input method developer first. It's the input method that needs to add workarounds for the gui toolkit, not the other way around. Let's recall a time when `ibus` couldn't input in Qt5 applications, which boosted the popularity of `fcitx` that could. Unikey is also patched into EVKey (closed source, though) to allow properly typing Vietnamese in the address bar of Chrome. Man, I'm foolish.
